orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1386 @@
+ *              the SAI attribute as well.
+ *
+ * Params:      IN  weight - The weight of the next hop group member.
+ *
+ * Returns:     true, if the operation was successful;
+ *              false, otherwise.
+ */
+bool NextHopGroupMember::updateWeight(uint32_t weight)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_INFO("Update group member %s weight from %u to %u",
+                    m_nh_key.to_string().c_str(),
+                    m_nh_key.weight,
+                    weight);
+
+    bool success = true;
+
+    m_nh_key.weight = weight;
+
+    if (isSynced())
+    {
+        SWSS_LOG_INFO("Updating SAI weight attribute");
+
+        sai_attribute_t nhgm_attr;
+        nhgm_attr.id = SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT;
+        nhgm_attr.value.s32 = m_nh_key.weight;
+
+        sai_status_t status = sai_next_hop_group_api->
+                set_next_hop_group_member_attribute(m_gm_id, &nhgm_attr);
+        success = status == SAI_STATUS_SUCCESS;
+    }
+
+    SWSS_LOG_INFO("Returning %d", success);
+    return success;
+}
+
+/*
+ * Purpose:     Sync the group member with the given group member ID.
+ *
+ * Description: Set the group member's SAI ID to the the one given and
+ *              increment the appropriate ref counters.
+ *
+ * Params:      IN  gm_id - The group member SAI ID to set.
+ *
+ * Returns:     Nothing.
+ */
+void NextHopGroupMember::sync(sai_object_id_t gm_id)
+{
+    SWSS_LOG_ENTER();
+
+    /* The SAI ID should be updated from invalid to something valid. */
+    assert((m_gm_id == SAI_NULL_OBJECT_ID) && (gm_id != SAI_NULL_OBJECT_ID));
+
+    m_gm_id = gm_id;
+    gCrmOrch->incCrmResUsedCounter(CrmResourceType::CRM_NEXTHOP_GROUP_MEMBER);
+    gNeighOrch->increaseNextHopRefCount(m_nh_key);
+}
+
+/*
+ * Purpose:     Desync the group member, resetting it's SAI ID.
+ *
+ * Description: Reset the group member's SAI ID and decrement the appropriate
+ *              ref counters.
+ *
+ * Params:      None.
+ *
+ * Returns:     Nothing.
+ */
+void NextHopGroupMember::desync()
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * If the member is already desynced, exit so we don't decrement the ref
+     * counters more than once.
+     */
+    if (!isSynced())
+    {
+        SWSS_LOG_INFO("Next hop group member %s is already desynced",
+                        m_nh_key.to_string().c_str());
+        return;
+    }
+
+    m_gm_id = SAI_NULL_OBJECT_ID;
+    gCrmOrch->decCrmResUsedCounter(CrmResourceType::CRM_NEXTHOP_GROUP_MEMBER);
+    gNeighOrch->decreaseNextHopRefCount(m_nh_key);
+}
+
+/*
+ * Purpose:     Destructor.
+ *
+ * Description: Assert the group member is desynced and remove the labeled
+ *              next hop from NeighOrch if it is unreferenced.
+ *
+ * Params:      None.
+ *
+ * Returns:     Nothing.
+ */
+NextHopGroupMember::~NextHopGroupMember()
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * The group member should be desynced from it's group before destroying
+     * it.
+     */
+    assert(!isSynced());
+
```

Comment:
Doesn't the same login apply to non-labeled nexthops? what happens when a Neighbor entry is created in gNeighOrch due to nhgOrch? How will that entry get deleted?

Reply:
Non-labeled next hops are only created by neighOrch based on a neighbor being created. Whereas labeled next hops are created when routeorch/labelrouteorch/nhgorch ask neighorch to. So nhgOrch will never cause a Neighbor entry to be created in gNeighOrch

Reply:
What happens when a next hop is added from nghOrch for which the nexthop does not have ARP resolved, and/or that nexthop is not used by routeorch.

Reply:
If a NHG is created with a next hop that isn't resolved nhgorch does the same thing as routeorch. It will fail to program the NHG, but keep it on the list to be retried. Any route that references the NHG will then also wait for the NHG to be ready.

I've just added a commit to make nhgorch as neighorch to resolve the next hop - I noticed this had been added to routeorch since our code for this was originally written.